### ISSUE TITLE
False positive map_clone on .map(|x| *x) on Option<Box<T>>

### ISSUE THREAD

**Issue by dtolnay** [rust-lang-nursery/rust-clippy#3274]
```rust
fn main() {
    let x = Some(Box::new(true));
    let y = x.map(|x| *x);
    println!("{:?}", y);
}
```

```console
warning: You are using an explicit closure for cloning elements                
 --> src/main.rs:3:13                                                          
  |                                                                            
3 |     let y = x.map(|x| *x);                                                 
  |             ^^^^^^^^^^^^^ help: Consider calling the dedicated `cloned` method: `x.cloned()`
  |                                                                            
  = note: #[warn(clippy::map_clone)] on by default                             
  = help: for further information visit https://rust-lang-nursery.github.io/rust-clippy/v0.0.212/index.html#map_clone
```

The suggested code does not compile.

```console
error[E0599]: no method named `cloned` found for type `std::option::Option<std::boxed::Box<bool>>` in the current scope
 --> src/main.rs:3:15                                                          
  |                                                                            
3 |     let y = x.cloned();                                                    
  |               ^^^^^^                                                       
  |                                                                            
  = note: the method `cloned` exists but the following trait bounds were not satisfied:
          `&mut std::option::Option<std::boxed::Box<bool>> : std::iter::Iterator`
```
